File: android/src/main/java/com/reveila/android/db/GenericDao.java
package com.reveila.android.db;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;

import java.util.List;

/**
 * Java Data Access Object for generic entity persistence.
 * Handles the low-level SQL operations for the Sovereign Engine.
 */
@Dao
public interface GenericDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insert(GenericEntity entity);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertAll(List<GenericEntity> entities);

    @Query("SELECT * FROM generic_entities WHERE id = :id AND type = :type LIMIT 1")
    GenericEntity findById(String id, String type);

    @Query("SELECT * FROM generic_entities WHERE type = :type")
    List<GenericEntity> findByType(String type);

    @Query("DELETE FROM generic_entities WHERE id = :id AND type = :type")
    void deleteById(String id, String type);

    @Query("SELECT COUNT(*) FROM generic_entities WHERE type = :type")
    long countByType(String type);
    
    @Query("SELECT COUNT(*) FROM generic_entities WHERE id = :id AND type = :type")
    int exists(String id, String type);
    
    /**
     * Simplistic pagination without sorting/filtering to satisfy the repository interface.
     */
    @Query("SELECT * FROM generic_entities WHERE type = :type LIMIT :limit OFFSET :offset")
    List<GenericEntity> fetchPage(String type, int limit, int offset);
}

File: android/src/main/kotlin/com/reveila/android/db/GenericDao.kt
package com.reveila.android.db

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query

@Dao
interface GenericDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insert(entity: GenericEntity)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertAll(entities: List<GenericEntity>)

    @Query("SELECT * FROM generic_entities WHERE id = :id AND type = :type LIMIT 1")
    fun findById(id: String, type: String): GenericEntity?

    @Query("SELECT * FROM generic_entities WHERE type = :type")
    fun findByType(type: String): List<GenericEntity>

    @Query("DELETE FROM generic_entities WHERE id = :id AND type = :type")
    fun deleteById(id: String, type: String)

    @Query("SELECT COUNT(*) FROM generic_entities WHERE type = :type")
    fun countByType(type: String): Long
    
    @Query("SELECT COUNT(*) FROM generic_entities WHERE id = :id AND type = :type")
    fun exists(id: String, type: String): Int
    
    // Simplistic pagination without sorting/filtering for now, 
    // to satisfy the interface.
    @Query("SELECT * FROM generic_entities WHERE type = :type LIMIT :limit OFFSET :offset")
    fun fetchPage(type: String, limit: Int, offset: Int): List<GenericEntity>
}
